docs/performance/optimization.rst

```diff
@@ -0,0 +1,598 @@
+     factory_metadata.factory_name,
+     device_data.device_name,
+     device_data.reading_value
+   FROM device_data
+   INNER JOIN factory_metadata ON device_data.factory_id = factory_metadata.factory_id
+   WHERE reading_time BETWEEN '2024-01-01' AND '2025-01-01'
+   LIMIT 10;
+
+do:
+
+.. code:: sql
+
+   WITH filtered_device_data AS (
+      SELECT
+            device_data.factory_id,
+            device_data.device_name,
+            device_data.reading_value
+      FROM device_data
+      WHERE
+           /*
+             We are sure one month of data is sufficient to find
+             10 results and it may help with partition pruning
+           */
+            reading_time BETWEEN '2024-12-01' AND '2025-01-01'
+      LIMIT 10
+   )
+   SELECT
+     factory_metadata.factory_name,
+     filtered_device_data.device_name,
+     filtered_device_data.reading_value
+   FROM filtered_device_data
+   INNER JOIN factory_metadata ON filtered_device_data.factory_id = factory_metadata.factory_id;
+
+.. _filter-with-array-expressions:
+
+Use filters with array expressions when filtering on the output of UNNEST
+=========================================================================
+
+On denormalized data sets you may have records with an array of objects.
+
+You may want to unnest the array in a subquery or CTE and later filter on a
+property of the OBJECTs.
+
+The next statement will result in every row in the table (not filtered
+with other conditions) being read and unnested, to check if it meets
+the criteria on ``field1``.
+
+.. code:: sql
+
+   SELECT *
+   FROM (
+      SELECT UNNEST(my_array_of_objects) obj
+      FROM my_table
+   )
+   WHERE obj['field1'] = 1;
+
+However, CrateDB can do a lot better than this if we add an additional condition
+like this:
+
+.. code:: sql
+
+   SELECT *
+   FROM (
+      SELECT UNNEST(my_array_of_objects) obj
+      FROM my_table
+      WHERE 1 = ANY(my_array_of_objects['field1'])
+   ) AS subquery
+   WHERE obj['field1'] = 1;
+
+CrateDB leverages indexes to only unnest the relevant records from ``my_table``
+which can make a huge difference.
+
+.. _group-efficient-query-structure:
+
+******************************************
+ Efficient Query Structure and Constructs
+******************************************
+
+This section focuses on optimizing SQL logic by prioritizing efficient syntax
+and avoiding redundant operations.
+
+.. _only-sort-when-needed:
+
+Only sort data when needed
+==========================
+
+Indexing in CrateDB is optimized to support filtering and aggregations without
+requiring expensive defragmentation operations, but it is not optimized for
+sorting​.
+
+Maintaining a sorted index would slow down ingestion, that is why​ other
+analytical database systems like Cassandra and Redshift make similar
+trade-offs​.
+
+This means that when an ``ORDER BY`` operation is requested, the whole dataset
+needs to be loaded into the main memory on the relevant cluster node to be sorted.
```

Comment:
This isn't correct as CrateDB will push down any ``ORDER BY` as much as possible, in best case using an optimized lucene collector. Additional, intermediated node based ordering is applied and results from nodes are merged by a sorted merge on the handler node.